File: examples/showcase/slack_fronting/custom-code-bundles/slack_fronting_typescript/tests/slack-live-smoke.test.ts
import { describe, expect, it } from "vitest";
import type { InvocationContext } from "@anip-dev/service";
import { backendAdapter } from "../src/runtime/backend-adapter.js";
import { generatedCapabilities } from "../src/generated/capabilities.js";
import { generatedCapabilityMetadata, type BackendInvocationPlan, type GeneratedCapabilityRuntimeMetadata } from "../src/generated/runtime-target.js";

const configured = Boolean(process.env.SLACK_BOT_TOKEN && process.env.SLACK_CHANNEL_ID);

function capability(id: string): GeneratedCapabilityRuntimeMetadata {
  const found = generatedCapabilityMetadata.find((item) => item.capability_id === id);
  if (!found) throw new Error(`Missing capability ${id}`);
  return found;
}

function plan(parameters: Record<string, unknown>): BackendInvocationPlan {
  return {
    selected_binding: null,
    semantic_input: parameters,
    adapter_input: parameters,
    backend_input_contract: { mode: "explicit", required: [], optional: [] },
    unresolved_required_backend_inputs: [],
  };
}

async function slackPost(path: string, body: Record<string, unknown>): Promise<Record<string, unknown>> {
  const form = new URLSearchParams();
  for (const [key, value] of Object.entries(body)) form.set(key, String(value));
  const response = await fetch(`https://slack.com/api/${path}`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${process.env.SLACK_BOT_TOKEN}`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: form,
  });
  expect(response.ok).toBe(true);
  return await response.json() as Record<string, unknown>;
}

describe.skipIf(!configured)("Slack live backend adapter", () => {
  it("executes bounded reads and prepares governed previews without mutation", async () => {
    const channelId = process.env.SLACK_CHANNEL_ID!;
    process.env.ANIP_SLACK_ALLOWED_CHANNELS ??= channelId;
    const history = await slackPost("conversations.history", { channel: channelId, limit: 1 });
    expect(history.ok).toBe(true);
    const messages = (history.messages ?? []) as Record<string, unknown>[];
    const threadTs = messages.length > 0 ? String(messages[0].thread_ts ?? messages[0].ts) : null;

    const context = await backendAdapter.execute(capability("slack.channel.read_context"), plan({ channel_id: channelId, limit: 5 }), { channel_id: channelId, limit: 5 }, {});
    expect(context.execution_status).toBe("completed");

    if (threadTs) {
      const thread = await backendAdapter.execute(capability("slack.thread.summarize"), plan({ channel_id: channelId, thread_ts: threadTs, limit: 10 }), { channel_id: channelId, thread_ts: threadTs, limit: 10 }, {});
      expect(thread.execution_status).toBe("completed");
    }

    const previews: Array<[string, Record<string, unknown>]> = [
      ["slack.message.prepare", { channel_id: channelId, text: "ANIP Slack TypeScript smoke preview" }],
      ["slack.incident_update.prepare", { channel_id: channelId, incident_id: "INC-123", status: "monitoring", summary: "Preview only", next_update_time: "in 30 minutes" }],
      ["slack.announcement.request", { channel_id: channelId, announcement: "Preview governed announcement only", audience: "internal" }],
    ];
    for (const [id, parameters] of previews) {
      const result = await backendAdapter.execute(capability(id), plan(parameters), parameters, {});
      expect(result.execution_status).toBe("prepared");
      expect(result.mutation_performed).toBe(false);
    }

    if (process.env.ANIP_SLACK_ALLOW_SEND === "true") {
      const sendParameters = { channel_id: channelId, text: `ANIP approved Slack TypeScript post at ${Date.now()}` };
      const sent = await backendAdapter.execute(capability("slack.message.prepare"), plan(sendParameters), sendParameters, { approvalGrant: "grant_live_typescript_smoke" });
      expect(sent.execution_status).toBe("completed");
      expect(sent.mutation_performed).toBe(true);
      expect((sent.posted_message as Record<string, unknown>).ts).toBeTruthy();
    }
  });

  it("routes approved sends through the generated handler", async () => {
    const channelId = process.env.SLACK_CHANNEL_ID!;
    process.env.ANIP_SLACK_ALLOWED_CHANNELS ??= channelId;
    const capabilityDef = generatedCapabilities.find((item) => item.declaration.name === "slack.announcement.request");
    expect(capabilityDef).toBeTruthy();
    const parameters = {
      channel_id: channelId,
      announcement: `ANIP approved Slack TypeScript generated handler post at ${Date.now()}`,
      audience: "test",
    };

    await expect(capabilityDef!.handler(testContext(null), parameters)).rejects.toMatchObject({
      errorType: "approval_required",
    });

    if (process.env.ANIP_SLACK_ALLOW_SEND === "true") {
      const sent = await capabilityDef!.handler(testContext("grant_live_typescript_handler_smoke"), parameters);
      expect(sent.execution_status).toBe("completed");
      expect(sent.mutation_performed).toBe(true);
      expect((sent.posted_message as Record<string, unknown>).ts).toBeTruthy();
    }
  });
});

function testContext(approvalGrant: string | null): InvocationContext {
  return {
    token: {} as InvocationContext["token"],
    rootPrincipal: "human:local-dev|actor_id=slack_requester",
    subject: "agent:slack-live-smoke",
    scopes: ["slack.announcement.request"],
    delegationChain: [],
    invocationId: "inv-test",
    clientReferenceId: null,
    taskId: null,
    parentInvocationId: null,
    upstreamService: null,
    approvalGrant,
    setCostActual: () => {},
    emitProgress: async () => {},
  };
}

File: examples/showcase/slack_fronting/custom-code-bundles/slack_fronting_go/extensions/slack_live_smoke_test.go
package extensions

import (
	"os"
	"testing"

	"{{ANIP_GO_MODULE_PATH}}/generated"
)

func TestSlackLiveBackendAdapter(t *testing.T) {
	token := textEnv("SLACK_BOT_TOKEN")
	channelID := textEnv("SLACK_CHANNEL_ID")
	if token == "" || channelID == "" {
		t.Skip("Slack credentials are not configured")
	}
	if textEnv("ANIP_SLACK_ALLOWED_CHANNELS") == "" {
		t.Setenv("ANIP_SLACK_ALLOWED_CHANNELS", channelID)
	}
	history := slackPost(token, "conversations.history", map[string]any{"channel": channelID, "limit": 1})
	if history["ok"] != true {
		t.Fatalf("Slack history failed: %#v", history)
	}
	var threadTS string
	if messages, ok := history["messages"].([]any); ok && len(messages) > 0 {
		message := messages[0].(map[string]any)
		threadTS = text(firstNonNil(message["thread_ts"], message["ts"]))
	}
	context := executeForTest(t, "slack.channel.read_context", map[string]any{"channel_id": channelID, "limit": 5})
	if context["execution_status"] != "completed" {
		t.Fatalf("expected channel read completed, got %#v", context)
	}
	if threadTS != "" {
		thread := executeForTest(t, "slack.thread.summarize", map[string]any{"channel_id": channelID, "thread_ts": threadTS, "limit": 10})
		if thread["execution_status"] != "completed" {
			t.Fatalf("expected thread read completed, got %#v", thread)
		}
	}
	cases := map[string]map[string]any{
		"slack.message.prepare": {"channel_id": channelID, "text": "ANIP Slack Go smoke preview"},
		"slack.incident_update.prepare": {"channel_id": channelID, "incident_id": "INC-123", "status": "monitoring", "summary": "Preview only", "next_update_time": "in 30 minutes"},
		"slack.announcement.request": {"channel_id": channelID, "announcement": "Preview governed announcement only", "audience": "internal"},
	}
	for id, params := range cases {
		result := executeForTest(t, id, params)
		if result["execution_status"] != "prepared" || result["mutation_performed"] != false {
			t.Fatalf("%s expected prepared without mutation, got %#v", id, result)
		}
	}
	if textEnv("ANIP_SLACK_ALLOW_SEND") == "true" {
		params := map[string]any{"channel_id": channelID, "text": "ANIP approved Slack Go post"}
		result := executeForTestWithContext(t, "slack.message.prepare", params, BackendInvocationContext{ApprovalGrant: "grant_live_go_smoke"})
		if result["execution_status"] != "completed" || result["mutation_performed"] != true {
			t.Fatalf("expected approved send completed, got %#v", result)
		}
		if posted, ok := result["posted_message"].(map[string]any); !ok || text(posted["ts"]) == "" {
			t.Fatalf("expected posted message timestamp, got %#v", result)
		}
	}
}

func executeForTest(t *testing.T, capabilityID string, params map[string]any) map[string]any {
	t.Helper()
	return executeForTestWithContext(t, capabilityID, params, BackendInvocationContext{})
}

func executeForTestWithContext(t *testing.T, capabilityID string, params map[string]any, context BackendInvocationContext) map[string]any {
	t.Helper()
	capability := capabilityForTest(t, capabilityID)
	plan := generated.BackendInvocationPlan{SemanticInput: params, AdapterInput: params, BackendInputContract: generated.EffectiveBackendInputContract{Mode: "explicit"}}
	result, err := BackendAdapterInstance.Execute(capability, plan, params, context)
	if err != nil {
		t.Fatal(err)
	}
	return result
}

func capabilityForTest(t *testing.T, capabilityID string) generated.GeneratedCapabilityRuntimeMetadata {
	t.Helper()
	for _, capability := range generated.GeneratedCapabilityMetadata {
		if capability.CapabilityID == capabilityID {
			return capability
		}
	}
	t.Fatalf("missing capability %s", capabilityID)
	return generated.GeneratedCapabilityRuntimeMetadata{}
}

func textEnv(name string) string {
	return text(getenv(name))
}

func getenv(name string) string {
	return text(os.Getenv(name))
}
